kvbc/tools/db_editor/include/kv_blockchain_db_editor.hpp

```diff
@@ -301,6 +301,93 @@ struct GetCategories {
   }
 };
 
+struct GetEarliestCategoryUpdates {
+  const bool read_only = true;
+  std::string description() const {
+    return "getEarliestCategoryUpdates CATEGORY-ID [BLOCK-VERSION-TO]\n"
+           "  Returns the latest blockID that contains the given category, started from BLOCK-VERSION-TO. If "
+           "BLOCK-VERSION-TO is not set, we start from the lastReachableBlock.";
+  }
+
+  std::string execute(const KeyValueBlockchain &adapter, const CommandArguments &args) const {
+    if (args.values.empty()) {
+      throw NotFoundException{"No Category ID was given"};
+    }
+    auto latestBlockID = adapter.getLastReachableBlockId();
+    if (args.values.size() >= 2) {
+      latestBlockID = toBlockId(args.values[1]);
+    }
+    auto cat = args.values.front();
+    std::map<std::string, std::string> cat_updates_map;
+    BlockId relevantBlockId = adapter.getGenesisBlockId();
+    for (auto block = adapter.getGenesisBlockId(); block <= latestBlockID; block++) {
+      auto updates = adapter.getBlockUpdates(block);
+      if (updates->categoryUpdates().kv.count(cat)) {
+        relevantBlockId = block;
+        cat_updates_map = getKVStr(updates->categoryUpdates().kv.at(cat));
+        break;
+      }
+    }
+    if (relevantBlockId == adapter.getGenesisBlockId()) {
+      if (!adapter.getBlockUpdates(relevantBlockId)->categoryUpdates().kv.count(cat)) {
+        throw NotFoundException{"Couldn't find category id in any block in the given range"};
+      }
+    }
+    std::map<std::string, std::string> out{
+        {"blockID", std::to_string(relevantBlockId)}, {"category", cat}, {"updates", toJson(cat_updates_map)}};
+    return toJson(out);
+  }
+};
+
+inline std::string getStaleKeysStr(std::vector<std::string> stale_keys) {
+  if (stale_keys.empty()) return std::string();
+  std::string strKeys;
+  strKeys += "[";
+  for (auto &k : stale_keys) {
+    strKeys += "\"" + concordUtils::bufferToHex(k.data(), k.size()) + "\"" + ",";
+  }
+  strKeys.erase(strKeys.size() - 1);
+  strKeys += "]";
+  return strKeys;
+}
+
+struct GetCategoryEarliestStale {
+  const bool read_only = true;
+  std::string description() const {
+    return "getCategoryEarliestStale CATEGORY-ID [BLOCK-VERSION-TO]\n"
+           "  Returns the latest blockID that contains the given category, started from BLOCK-VERSION-TO. If "
+           "BLOCK-VERSION-TO is not set, we start from the lastReachableBlock.";
```

Comment:
I think the description is incorrect.

Moreover, I guess we can call this command `GetCategoryStaleKeys` or similar. Also, the block parameter can be called just `BLOCK-ID` or `AS-OF-BLOCK-ID` as we want these stale keys per block.

Reply:
I think I've misunderstood the idea behind that command. It is basically looking for the first block the category has at least one stale key? Could you please explain its use case?

Reply:
discussed offline